### PR TITLE
port missing System.Memory benchmarks

### LINKED ISSUE #360: Port missing System.Memory benchmarks from CoreFX
Not all of the System.Memory benchmarks have been ported, but they have to be
https://github.com/dotnet/corefx/tree/master/src/System.Memory/tests/Performance

### PR DESCRIPTION
**Pull request by adamsitnik**
@ahsonkhan I went through all System.Memory benchmarks in CoreFX and these were missing.

I am going to send a separate PR for the System.Buffers benchmarks

Fixes #360

### CODE REVIEW

File: src/benchmarks/micro/corefx/System.Memory/Span.cs [hunk 1]
```diff
@@ -55,20 +56,30 @@ public void Setup()
...
-        
+
+        [Benchmark]
+        public int IndexOfAnyTwoValues() => new System.Span<T>(_emptyWithSingleValue).IndexOfAny(_notDefaultValue, _notDefaultValue);
+
+        [Benchmark]
+        public int IndexOfAnyThreeValues() => new System.Span<T>(_emptyWithSingleValue).IndexOfAny(_notDefaultValue, _notDefaultValue, _notDefaultValue);
+
+        [Benchmark]
+        public int IndexOfFourValues() => new System.Span<T>(_emptyWithSingleValue).IndexOfAny(new ReadOnlySpan<T>(_fourValues));
```
Comment (adamsitnik) [inline src/benchmarks/micro/corefx/System.Memory/Span.cs]:
```suggestion
        public int IndexOfAnyFourValues() => new System.Span<T>(_emptyWithSingleValue).IndexOfAny(new ReadOnlySpan<T>(_fourValues));
```

File: src/benchmarks/micro/corefx/System.Memory/Span.cs [hunk 2]
```diff
@@ -55,20 +56,30 @@ public void Setup()
         [Benchmark]
         public bool EndsWith() => new System.Span<T>(_array).EndsWith(new System.ReadOnlySpan<T>(_same).Slice(start: Size / 2));
 
-        [GlobalSetup(Targets = new [] { nameof(IndexOfValue), nameof(LastIndexOfValue), nameof(LastIndexOfAnyValues) })]
+        [GlobalSetup(Targets = new [] { nameof(IndexOfValue), nameof(LastIndexOfValue), nameof(LastIndexOfAnyValues), nameof(IndexOfFourValues), nameof(IndexOfAnyTwoValues), nameof(IndexOfAnyThreeValues) })]
```
Comment (adamsitnik) [inline src/benchmarks/micro/corefx/System.Memory/Span.cs]:
```suggestion
        [GlobalSetup(Targets = new [] { nameof(IndexOfValue), nameof(LastIndexOfValue), nameof(LastIndexOfAnyValues), nameof(IndexOfAnyFourValues), nameof(IndexOfAnyTwoValues), nameof(IndexOfAnyThreeValues) })]
```

File: src/benchmarks/micro/corefx/System.Collections/Contains/ContainsFalse.cs [hunk 3]
```diff
@@ -64,7 +65,21 @@ public bool Array()
...
-        
+
+#if NETCOREAPP3_0
+        [BenchmarkCategory(Categories.Span)]
+        [Benchmark]
+        public bool Span()
+        {
+            bool result = default;
+            Span<T> collection = new Span<T>(_array);
+            var notFound = _notFound;
```
Comment (ahsonkhan) [inline src/benchmarks/micro/corefx/System.Collections/Contains/ContainsFalse.cs]: nit: avoid var

File: src/benchmarks/micro/corefx/System.Memory/Slice.cs [hunk 4]
```diff
@@ -152,7 +152,75 @@ public System.ReadOnlyMemory<T> ReadOnlyMemoryStartLength()
...
             return memory;
         }
-        
+
+        [Benchmark(OperationsPerInvoke = 16)]
+        public System.Memory<T> MemorySpanStart()
+        {
+            System.Memory<T> memory = new System.Memory<T>(_nonEmptyArray);
+
+            Consume(memory.Span.Slice(Size / 2)); Consume(memory.Span.Slice(Size / 2));
```
Comment (ahsonkhan) [inline src/benchmarks/micro/corefx/System.Memory/Slice.cs]: Part of this test (initially) was to compare mem.Slice.Span and mem.Span.Slice. Can we retain that?
https://github.com/dotnet/corefx/blob/fd88c5df4bde4ef56522a1cd81da1ff69c92edc4/src/System.Memory/tests/Performance/Perf.Memory.Slice.cs#L45-L71

Or is this test porting something else?

**Comment by adamsitnik** [comment]
> Part of this test (initially) was to compare mem.Slice.Span and mem.Span.Slice. Can we retain that?

Sure! I added them back. The resuts:

```cmd
dotnet run -c Release -f netcoreapp3.0 --filter *Slice* --join
```

|          Type |                        Method |     Mean |     Error |
|-------------- |------------------------------ |---------:|----------:|
|   Slice<Byte> |                     SpanStart | 1.776 ns | 0.0121 ns |
| Slice<String> |                     SpanStart | 5.197 ns | 0.0378 ns |
|   Slice<Byte> |               SpanStartLength | 1.548 ns | 0.0329 ns |
| Slice<String> |               SpanStartLength | 5.283 ns | 0.1174 ns |
|   Slice<Byte> |             ReadOnlySpanStart | 1.726 ns | 0.0398 ns |
| Slice<String> |             ReadOnlySpanStart | 4.375 ns | 0.1334 ns |
|   Slice<Byte> |       ReadOnlySpanStartLength | 1.577 ns | 0.0362 ns |
| Slice<String> |       ReadOnlySpanStartLength | 4.469 ns | 0.0298 ns |
|   Slice<Byte> |                   MemoryStart | 1.671 ns | 0.0167 ns |
| Slice<String> |                   MemoryStart | 6.869 ns | 0.0332 ns |
|   Slice<Byte> |               MemoryStartSpan | 4.127 ns | 0.0814 ns |
| Slice<String> |               MemoryStartSpan | 9.428 ns | 0.0441 ns |
|   Slice<Byte> |             MemoryStartLength | 1.730 ns | 0.0913 ns |
| Slice<String> |             MemoryStartLength | 7.072 ns | 0.0687 ns |
|   Slice<Byte> |         MemoryStartLengthSpan | 3.833 ns | 0.0389 ns |
| Slice<String> |         MemoryStartLengthSpan | 9.692 ns | 0.0898 ns |
|   Slice<Byte> |           ReadOnlyMemoryStart | 1.671 ns | 0.0181 ns |
| Slice<String> |           ReadOnlyMemoryStart | 5.459 ns | 0.0589 ns |
|   Slice<Byte> |       ReadOnlyMemoryStartSpan | 4.145 ns | 0.0752 ns |
| Slice<String> |       ReadOnlyMemoryStartSpan | 9.014 ns | 0.0468 ns |
|   Slice<Byte> |     ReadOnlyMemoryStartLength | 1.696 ns | 0.0196 ns |
| Slice<String> |     ReadOnlyMemoryStartLength | 5.919 ns | 0.1144 ns |
|   Slice<Byte> | ReadOnlyMemoryStartLengthSpan | 3.872 ns | 0.0402 ns |
| Slice<String> | ReadOnlyMemoryStartLengthSpan | 9.086 ns | 0.0426 ns |
|   Slice<Byte> |               MemorySpanStart | 4.119 ns | 0.1175 ns |
| Slice<String> |               MemorySpanStart | 8.298 ns | 0.0385 ns |
|   Slice<Byte> |         MemorySpanStartLength | 4.165 ns | 0.1345 ns |
| Slice<String> |         MemorySpanStartLength | 8.394 ns | 0.0330 ns |
|   Slice<Byte> |       ReadOnlyMemorySpanStart | 4.140 ns | 0.0791 ns |
| Slice<String> |       ReadOnlyMemorySpanStart | 7.671 ns | 0.0315 ns |
|   Slice<Byte> | ReadOnlyMemorySpanStartLength | 4.079 ns | 0.1209 ns |
| Slice<String> | ReadOnlyMemorySpanStartLength | 7.685 ns | 0.0757 ns |

**Pull request merged**